File: WebCaster/WebCaster/Services/Casting/CastingManager.swift
import Foundation
import Combine

final class CastingManager: ObservableObject {
    static let shared = CastingManager()

    @Published var discoveredDevices: [CastDevice] = []
    @Published var connectedDevice: CastDevice?
    @Published var isDiscovering = false
    @Published var isCasting = false
    @Published var castError: String?

    private let dlnaService = DLNAService.shared
    private let chromecastService = ChromecastService.shared
    private let persistence = PersistenceController.shared
    private let discoveryGroup = DispatchGroup()

    private init() {
        loadRecentDevices()
    }

    func startDiscovery() {
        guard !isDiscovering else { return }
        isDiscovering = true
        castError = nil
        discoveredDevices.removeAll()

        var allDevices: [CastDevice] = []
        let lock = NSLock()

        discoveryGroup.enter()
        dlnaService.discoverDevices(timeout: 8) { devices in
            lock.lock()
            allDevices.append(contentsOf: devices)
            lock.unlock()
            self.discoveryGroup.leave()
        }

        discoveryGroup.enter()
        chromecastService.discoverDevices(timeout: 6) { devices in
            lock.lock()
            allDevices.append(contentsOf: devices)
            lock.unlock()
            self.discoveryGroup.leave()
        }

        discoveryGroup.notify(queue: .main) { [weak self] in
            self?.discoveredDevices = allDevices
            self?.isDiscovering = false
            self?.mergeWithRecentDevices()
        }
    }

    func stopDiscovery() {
        dlnaService.stopDiscovery()
        chromecastService.stopDiscovery()
        isDiscovering = false
    }

    func connect(to device: CastDevice) {
        var updatedDevice = device
        updatedDevice.state = .connected
        updatedDevice.lastConnected = Date()
        connectedDevice = updatedDevice
        persistence.saveRecentDevice(updatedDevice)

        if let index = discoveredDevices.firstIndex(where: { $0 == device }) {
            discoveredDevices[index] = updatedDevice
        }
    }

    func disconnect() {
        if let device = connectedDevice {
            dlnaService.stop(device: device) { _, _ in }
        }
        connectedDevice?.state = .available
        connectedDevice = nil
        isCasting = false
    }

    func cast(video: DetectedVideo) {
        guard let device = connectedDevice else {
            castError = "No device connected"
            return
        }

        isCasting = true
        castError = nil

        switch device.castProtocol {
        case .dlna:
            dlnaService.cast(video: video, to: device) { [weak self] success, error in
                DispatchQueue.main.async {
                    if success {
                        self?.connectedDevice?.state = .playing
                    } else {
                        self?.castError = error ?? "Failed to cast"
                        self?.isCasting = false
                    }
                }
            }
        case .airplay:
            break
        case .chromecast:
            chromecastService.launchMedia(on: device, videoURL: video.url) { [weak self] success, error in
                DispatchQueue.main.async {
                    if success {
                        self?.connectedDevice?.state = .playing
                    } else {
                        self?.castError = error ?? "Chromecast casting failed. For full support, integrate google-cast-sdk."
                        self?.isCasting = false
                    }
                }
            }
        }
    }

    func pauseCasting() {
        guard let device = connectedDevice, device.castProtocol == .dlna else { return }
        dlnaService.pause(device: device) { [weak self] success, _ in
            if success {
                DispatchQueue.main.async {
                    self?.connectedDevice?.state = .paused
                }
            }
        }
    }

    func resumeCasting() {
        guard let device = connectedDevice, device.castProtocol == .dlna else { return }
        dlnaService.play(device: device) { [weak self] success, _ in
            if success {
                DispatchQueue.main.async {
                    self?.connectedDevice?.state = .playing
                }
            }
        }
    }

    func stopCasting() {
        guard let device = connectedDevice, device.castProtocol == .dlna else { return }
        dlnaService.stop(device: device) { [weak self] _, _ in
            DispatchQueue.main.async {
                self?.connectedDevice?.state = .connected
                self?.isCasting = false
            }
        }
    }

    var devicesByProtocol: [CastProtocol: [CastDevice]] {
        Dictionary(grouping: discoveredDevices, by: \.castProtocol)
    }

    private func loadRecentDevices() {
        let recent = persistence.loadRecentDevices()
        discoveredDevices = recent
    }

    private func mergeWithRecentDevices() {
        let recent = persistence.loadRecentDevices()
        for device in recent {
            if !discoveredDevices.contains(device) {
                var stale = device
                stale.state = .available
                discoveredDevices.append(stale)
            }
        }
    }
}

File: WebCaster/src/services/casting/CastingManager.ts
import { CastDevice, DetectedVideo } from '../../models/types';
import { DLNA } from './DLNAService';
import { Storage } from '../storage/PersistenceController';

export type CastingState = {
  devices: CastDevice[];
  connectedDevice: CastDevice | null;
  isDiscovering: boolean;
  isCasting: boolean;
  error: string | null;
};

const initialState: CastingState = {
  devices: [],
  connectedDevice: null,
  isDiscovering: false,
  isCasting: false,
  error: null,
};

let state = { ...initialState };
let listeners: ((s: CastingState) => void)[] = [];

function notify() {
  listeners.forEach(fn => fn({ ...state }));
}

export const CastingManager = {
  subscribe(fn: (s: CastingState) => void) {
    listeners.push(fn);
    fn({ ...state });
    return () => { listeners = listeners.filter(l => l !== fn); };
  },

  getState(): CastingState {
    return { ...state };
  },

  async startDiscovery() {
    if (state.isDiscovering) return;
    state = { ...state, isDiscovering: true, error: null, devices: [] };
    notify();

    const recent = await Storage.loadRecentDevices();
    state = { ...state, devices: recent, isDiscovering: false };
    notify();
  },

  async connect(device: CastDevice) {
    const updated: CastDevice = { ...device, state: 'connected', lastConnected: Date.now() };
    state = { ...state, connectedDevice: updated };
    await Storage.saveRecentDevice(updated);
    notify();
  },

  disconnect() {
    if (state.connectedDevice) {
      if (state.connectedDevice.protocol === 'dlna') {
        DLNA.stop(state.connectedDevice);
      }
    }
    state = { ...state, connectedDevice: null, isCasting: false };
    notify();
  },

  async cast(video: DetectedVideo) {
    const device = state.connectedDevice;
    if (!device) {
      state = { ...state, error: 'No device connected' };
      notify();
      return;
    }

    state = { ...state, isCasting: true, error: null };
    notify();

    let success = false;
    switch (device.protocol) {
      case 'dlna':
        success = await DLNA.cast(video, device);
        break;
      case 'chromecast':
        state = { ...state, error: 'Chromecast requires react-native-google-cast native module' };
        notify();
        return;
      case 'airplay':
        break;
    }

    if (success) {
      state = { ...state, connectedDevice: { ...device, state: 'playing' } };
    } else {
      state = { ...state, error: 'Failed to cast video', isCasting: false };
    }
    notify();
  },

  async pauseCasting() {
    const device = state.connectedDevice;
    if (!device) return;
    if (device.protocol === 'dlna') {
      const ok = await DLNA.pause(device);
      if (ok) state = { ...state, connectedDevice: { ...device, state: 'paused' } };
    }
    notify();
  },

  async resumeCasting() {
    const device = state.connectedDevice;
    if (!device) return;
    if (device.protocol === 'dlna') {
      const ok = await DLNA.play(device);
      if (ok) state = { ...state, connectedDevice: { ...device, state: 'playing' } };
    }
    notify();
  },

  async stopCasting() {
    const device = state.connectedDevice;
    if (!device) return;
    if (device.protocol === 'dlna') await DLNA.stop(device);
    state = { ...state, connectedDevice: { ...device, state: 'connected' }, isCasting: false };
    notify();
  },

  clearError() {
    state = { ...state, error: null };
    notify();
  },
};
